src/query/server/server.go

```diff
@@ -313,38 +362,87 @@ func newDownsampler(
 		TagDecoderPoolOptions: tagDecoderPoolOptions,
 	})
 	if err != nil {
-		logger.Fatal("unable to create downsampler", zap.Any("error", err))
+		return nil, errors.Wrap(err, "unable to create downsampler")
+	}
+
+	return downsampler, nil
+}
+
+func initClusters(cfg config.Configuration, dbClientCh <-chan client.Client, logger *zap.Logger) (local.Clusters, error) {
+	var (
+		clusters local.Clusters
+		err      error
+	)
+
+	if len(cfg.Clusters) > 0 {
+		opts := local.ClustersStaticConfigurationOptions{
+			AsyncSessions: true,
+		}
+		clusters, err = cfg.Clusters.NewClusters(opts)
+		if err != nil {
+			return nil, errors.Wrap(err, "unable to connect to clusters")
+		}
+	} else {
+		localCfg := cfg.Local
+		if localCfg == nil {
+			localCfg = defaultLocalConfiguration
+		}
+
+		if dbClientCh == nil {
+			return nil, errors.New("no clusters configured and not running local cluster")
+		}
+		session := m3db.NewAsyncSession(func() (client.Client, error) {
+			return <-dbClientCh, nil
+		}, nil)
+		clusters, err = local.NewClusters(local.UnaggregatedClusterNamespaceDefinition{
+			NamespaceID: ident.StringID(localCfg.Namespace),
+			Session:     session,
+			Retention:   localCfg.Retention,
+		})
+		if err != nil {
+			return nil, errors.Wrap(err, "unable to connect to clusters")
+		}
 	}
 
-	return downsampler
+	for _, namespace := range clusters.ClusterNamespaces() {
+		logger.Info("resolved cluster namespace",
+			zap.String("namespace", namespace.NamespaceID().String()))
+	}
+
+	return clusters, nil
 }
 
 func newStorages(
 	logger *zap.Logger,
 	clusters local.Clusters,
 	cfg config.Configuration,
 	workerPool pool.ObjectPool,
-) (storage.Storage, func()) {
-	cleanup := func() {}
+) (storage.Storage, cleanupFn, error) {
+	cleanup := func() error { return nil }
 
 	localStorage := local.NewStorage(clusters, workerPool)
 	stores := []storage.Storage{localStorage}
 	remoteEnabled := false
 	if cfg.RPC != nil && cfg.RPC.Enabled {
 		logger.Info("rpc enabled")
-		server := startGrpcServer(logger, localStorage, cfg.RPC)
-		cleanup = func() {
+		server, err := startGrpcServer(logger, localStorage, cfg.RPC)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		cleanup = func() error {
 			server.GracefulStop()
+			return nil
 		}
 
-		if remotes := cfg.RPC.RemoteListenAddresses; len(remotes) > 0 {
-			client, err := tsdbRemote.NewGrpcClient(remotes)
-			if err != nil {
-				logger.Fatal("unable to start remote clients for addresses", zap.Any("error", err))
-			}
+		remoteStorage, enabled, err := remoteClient(cfg)
+		if err != nil {
+			return nil, nil, err
```

Comment:
Worth logging?

Reply:
trying to avoid error/fatal logging within functions and only do that within Run or go routines.